walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,108 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 
+/**
+ * This class only contains the query used to find the time ranges when a
+ * meeting request coincides with available time.
+ */
 public final class FindMeetingQuery {
+  // This comparator is used to sort a list of events by start time
+  // in ascending order.
+  private Comparator<Event> eventComparator = new Comparator<Event>() {
+    @Override
+    public int compare(Event event1, Event event2) {
+      return event1.getWhen().start() - event2.getWhen().start();
+    }
+  };
+  /**
+   * Given a set of events and a meeting request, the query returns a list of
+   * free time ranges available for the meeting to be scheduled. If no valid
+   * time range is found, returns an empty list.
+   *
+   * Time complexity: 
+   * n: number of events
+   * m: number of attendees in every event
+   * O(nm)
+   *
+   * Space complexity:
+   * n: number of events
+   * m: number of available time ranges
+   * O(m+n)
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // We pass the events into a list since the collection can't be sorted,
+    // and we sort the list by event start time in ascending order.
+    List<Event> eventList = new ArrayList(events);
+    Collections.sort(eventList, eventComparator);
+
+    // Since there are no previous events yet, the start of day is also the
+    // start of the first range to check available time.
+    int pastEventEnd = TimeRange.START_OF_DAY;
+    int currentEventStart;
+    int availableTimeDuration;
+    List<TimeRange> availableTimes = new ArrayList<>();
+    for (Event event : eventList) {
+      if (!eventSharesAttendeesWithRequest(event, request)) {
+        // If the event does not share attendes with the request, it means that
+        // it does not change time availability, so it's skipped.
+        continue;
+      }
+      currentEventStart = event.getWhen().start();
+      availableTimeDuration = currentEventStart - pastEventEnd;
+
```

Comment:
nit: probably can remove this empty line. If anything, put an empty line before 'currentEventStart =...', because that line and the following if statements all are directly related logic to each other.

Reply:
Done